src/ec/Jacobi.mo

```diff
@@ -0,0 +1,420 @@
+        };
+      };
+      case _ {
+        false;
+      };
+    };
+  };
+
+  // Check if the given point is the point at infinity.
+  public func isInfinity(point : Point) : Bool {
+    return switch point {
+      case (#infinity (_)) {
+        true;
+      };
+      case (#point (_, _, z, curve)) {
+        z.isEqual(curve.Fp(0));
+      }
+    };
+  };
+
+  // Convert the given jacobi point to affine.
+  public func toAffine(point: Point) : Affine.Point {
+    let scaledPoint = scale(point);
+    return switch scaledPoint {
+      case (#infinity (curve)) {
+        #infinity (curve)
+      };
+      case (#point (x, y, _, curve)) {
+        #point (x, y, curve)
+      };
+    };
+  };
+
+  // Invert the given point on the x-axis.
+  public func neg(point: Point) : Point {
+    return switch (normalizeInfinity(point)) {
+      case (#infinity (curve)) {
+        #infinity (curve)
+      };
+      case (#point (x, y, z, curve)) {
+        #point (x, y.neg(), z, curve)
+      };
+    }
+  };
+
+  // Normalize the given point such that z = 1.
+  public func scale(point: Point): Point {
+    switch (normalizeInfinity(point)) {
+      case (#infinity (curve)) {
+        return #infinity (curve);
+      };
+      case (#point (x, y, z, curve)) {
+        if (z.isEqual(curve.Fp(1))) {
+          return point;
+        };
+
+        let zInverse = z.inverse();
+        let zzInverse = zInverse.sqr();
+
+        let newX = x.mul(zzInverse);
+        let newY = y.mul(zzInverse).mul(zInverse);
+        let newZ = curve.Fp(1);
+
+        return #point (newX, newY, newZ, curve);
+      };
+    };
+  };
+
+  // Return double of the given point.
+  public func double(point : Point) : Point {
+    switch (normalizeInfinity(point)) {
+      case (#infinity (curve)) {
+        return #infinity (curve);
+      };
+      case (#point (x, y, z, curve)) {
+        let (x2, y2, z2) = doDouble(x.value, y.value, z.value, curve.a, curve.p);
+        return #point (fpFromInt(x2, curve), fpFromInt(y2, curve),
+            fpFromInt(z2, curve), curve);
+      };
+    };
+  };
+
+  // Multiply the given point by the given scalar value.
+  public func mul(point: Point, other: Nat) : Point {
+    if (other == 0) {
+      return #infinity (getCurve(point));
+    };
+    switch (scale(point)) {
+      case (#infinity (curve)) {
+        return #infinity (curve);
+      };
+      case (#point (x2, y2, _, curve)) {
+        var p : (Int, Int, Int) = (0, 0, 1);
+        let naf = Numbers.toNaf(other);
+        let y2neg = y2.neg().value;
+
+        for (i in Iter.revRange(naf.size() - 1, 0)) {
+          let nafItem : Int = naf[Int.abs(i)];
+          p := doDouble(p.0, p.1, p.2, curve.a, curve.p);
+          if (nafItem < 0) {
+            p := _add(p.0, p.1, p.2, x2.value, y2neg, 1,
+              curve.a, curve.p);
+          } else if (nafItem > 0) {
+            p := _add(p.0, p.1, p.2, x2.value, y2.value, 1, curve.a, curve.p);
+          };
```

Comment:
Probably the biggest wins will come from optimizing the scalar multiplication. NAF is a good start, but consider using a windowed NAF instead.

Also, for ECDSA verification, multiexponentiations can provide notable improvements. The most basic form is what is known as "Shamir's Trick". There are some more complicated versions in https://www.bmoeller.de/pdf/multiexp-sac2001.pdf